### ISSUE TITLE
Startup is very slow

### ISSUE THREAD

**Issue by DroneKid** [airblade/vim-gitgutter#599]
Hi!

The startup is very slow on the latest master. It is blazingly fast if I roll back to `932ffaca092cca246b82c33e23d2d3a05e192e08`. It looks like `5bfe5b92099e0d332a7ca26c0263fab509ee9617 Big refactor` had introduced something to slow it down.

Slow startup is only experrienced if I pass a file to vim at command line.

> What is the latest commit SHA in your installed vim-gitgutter?
`7eeea63e62b1cc088a75c7a7c244fc774d82e5bb`

> What vim/nvim version are you on?
vim 8.1.1000

Cheers,
A.

**Comment by airblade** [comment]
I can't reproduce this.  Please could you [profile](https://github.com/airblade/vim-gitgutter/wiki/How-to-profile-vim-gitgutter) it?

**Comment by Morozzzko** [comment]
I'm having a similar problem. 

I'm at `7eeea63e62b1cc088a75c7a7c244fc774d82e5bb`
NVIM v0.3.4

I'm going to attach the problematic parts of the log:

```
SCRIPT  /Users/moroz/.vim/plugged/vim-gitgutter/autoload/gitgutter/diff.vim
Sourced 1 time
Total time:   3.610975
 Self time:   0.000804
```

Here's the gist with profiling for `diff.vim`

https://gist.github.com/Morozzzko/5a1792f2cbc3ef0e12c845c66bc15e9e#file-gistfile1-txt-L17

**Comment by blueyed** [comment]
@Morozzzko 
So it seems the git call itself is slow (https://github.com/nedbat/coveragepy/blob/820b255f34a0aac8670b0c819153bb8b38c4b2c6/coverage/ctracer/tracer.c#L1015-L1017).  Can you run it manually?  Is it due to your shell startup being slow maybe etc?

**Comment by Morozzzko** [comment]
Right, thank you! That's what happened. Shell startup was slow, so I optimized it and everything became better. Thank you!

Sorry for hijacking the issue by the way. Seems like I've got a different kind of problem

**Comment by nicksergeant** [comment]
I'm also seeing about a .5s delay upon opening any new file into a new buffer, and I've narrowed it down to `vim-gitgutter`. I'm using NERDtree to open a file, and the problem only exists for new buffers. Once a file has been opened in the current Vim session, opening the same file is very fast (under 100ms, I'd guess).

I've profiled opening a few new files here: https://snipt.net/raw/95aa91d44419e095d1e4d908a0af5bf3/

**Comment by airblade** [comment]
@nicksergeant Thanks for the profile; it really helps.

The first thing the plugin does when you enter a buffer is [determine the path](https://github.com/airblade/vim-gitgutter/blob/7eeea63e62b1cc088a75c7a7c244fc774d82e5bb/autoload/gitgutter/utility.vim#L117-L153) from the repo root to the file by running a git command.  It does this asynchronously to avoid slowing down the UI.  Once it knows the path it caches it in a buffer variable.

The second thing it does is run a diff.  To do this it needs the path from the first step, but as the first step is asynchronous it needs to [poll for the result](https://github.com/airblade/vim-gitgutter/blob/7eeea63e62b1cc088a75c7a7c244fc774d82e5bb/autoload/gitgutter/diff.vim#L71-L73).

Your profile shows that the plugin is polling for ages:

```
FUNCTION  gitgutter#diff#run_diff()
    Defined: ~/.vim/plugged/vim-gitgutter/autoload/gitgutter/diff.vim line 70
Called 6 times
Total time:   1.159836
 Self time:   1.051948

count  total (s)   self (s)
  169   0.017994   0.005367   while gitgutter#utility#repo_path(a:bufnr, 0) == -1
  163   1.051244   1.038190     sleep 5m
  169              0.002272   endwhile
```

So the question is, why is the async command taking so long?  Maybe spawning a shell takes a while for you, as for @Morozzzko above.  Or, if you use neovim, you could be encountering the neovim bug described in #589 – in which case upgrading neovim should fix it.

(On this topic, there's a pull request (#593) to remove the polling altogether but I (still!) haven't looked at it properly yet.)

**Comment by blueyed** [comment]
Yeah, came here to promote https://github.com/airblade/vim-gitgutter/pull/593 - maybe you could give it a try here?
Until then it would also make sense to increase the sleep here to e.g. `50m` as a short term solution until it gets removed.

**Comment by airblade** [comment]
This should be fixed in gitgutter as of 064a3d6.

**Comment by airblade** [comment]
Closing because I think this is fixed.  Let me know if not.
